samples/extensions/open_cl_interop/open_cl_interop.cpp

```diff
@@ -0,0 +1,622 @@
+/* Copyright (c) 2021, Arm Limited and Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "open_cl_interop.h"
+
+#include "common/vk_common.h"
+#include "gui.h"
+#include "platform/filesystem.h"
+#include "platform/platform.h"
+
+#define CL_FUNCTION_DEFINITIONS
+#include <open_cl_utils.h>
+
+struct CLData
+{
+	cl_context       context{nullptr};
+	cl_device_id     device_id{nullptr};
+	cl_command_queue command_queue{nullptr};
+	cl_program       program{nullptr};
+	cl_kernel        kernel{nullptr};
+	cl_mem           image{nullptr};
+};
+
+OpenCLInterop::OpenCLInterop()
+{
+	zoom  = -3.5f;
+	title = "Interoperability with OpenCL";
+
+	add_device_extension(VK_ANDROID_EXTERNAL_MEMORY_ANDROID_HARDWARE_BUFFER_EXTENSION_NAME);
+
+	add_device_extension(VK_KHR_SAMPLER_YCBCR_CONVERSION_EXTENSION_NAME);
+	add_device_extension(VK_KHR_MAINTENANCE1_EXTENSION_NAME);
+	add_device_extension(VK_KHR_BIND_MEMORY_2_EXTENSION_NAME);
+	add_device_extension(VK_KHR_GET_MEMORY_REQUIREMENTS_2_EXTENSION_NAME);
+	add_instance_extension(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+	add_instance_extension(VK_KHR_EXTERNAL_MEMORY_CAPABILITIES_EXTENSION_NAME);
+	add_device_extension(VK_KHR_EXTERNAL_MEMORY_EXTENSION_NAME);
+	add_device_extension(VK_EXT_QUEUE_FAMILY_FOREIGN_EXTENSION_NAME);
+	add_device_extension(VK_KHR_DEDICATED_ALLOCATION_EXTENSION_NAME);
+}
+
+OpenCLInterop::~OpenCLInterop()
+{
+	device->wait_idle();
+
+	vkDestroyPipeline(device->get_handle(), pipeline, nullptr);
+	vkDestroyPipelineLayout(device->get_handle(), pipeline_layout, nullptr);
+	vkDestroyDescriptorSetLayout(device->get_handle(), descriptor_set_layout, nullptr);
+	vkDestroyFence(device->get_handle(), rendering_finished_fence, nullptr);
+	vkDestroySampler(device->get_handle(), shared_texture.sampler, nullptr);
+	vkDestroyImageView(device->get_handle(), shared_texture.view, nullptr);
+	vkDestroyImage(device->get_handle(), shared_texture.image, nullptr);
+	vkFreeMemory(device->get_handle(), shared_texture.memory, nullptr);
+
+	clReleaseMemObject(cl_data->image);
+	clReleaseContext(cl_data->context);
+	delete cl_data;

```

Comment:
recommend testing that cl_data is not nullptr as its creation is in prepare and not the initializer (in other words even on supported platforms, this *could* cause a crash bug on exit).